pkg/storage/watch_cache.go

```diff
@@ -250,6 +250,20 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]wa
 	if size > 0 {
 		oldest = w.cache[w.startIndex%w.capacity].resourceVersion
 	}
+	if resourceVersion == 0 {
+		// resourceVersion = 0 means that we don't require any specific starting point
+		// and we would like to start watching from ~now.
+		// However, to keep backward compatibility, we additionally need to return the
+		// current state and only then start watching from that point.
+		//
+		// TODO: In v2 api, we should stop returning the current state - #13969.
+		allItems := w.store.List()
```

Comment:
I suspect he is referring to a potential race between the backing cache.Store and the watch cache.  So if the current items are enumerated from the cache.store, and a watch is initiated, its possible that in the time between that a change in etcd was missed.  I think to fix that you would need to know the resource version of the most recent item for when ```w.store.List()``` returned its value.